### PR TITLE
Check that events are at least expected ones

### LINKED ISSUE #3374: Intermittent Cloud Event failures during integration tests
I have noticed this kind of failure happening in many PRs recently. 

* [3369](https://tekton-releases.appspot.com/build/tekton-prow/pr-logs/pull/tektoncd_pipeline/3369/pull-tekton-pipeline-integration-tests/1315726113254674432/)
* [3353](https://github.com/tektoncd/pipeline/pull/3353#issuecomment-705080791)
* [3242](https://github.com/tektoncd/pipeline/pull/3242)

/cc @afrittoli 

```
TestTaskRunPipelineRunCancel/retries=0: cancel_test.go:159: Expected 2 number of successful events from pipelinerun and taskrun but got 3; list of received events : "&v1.Event{TypeMeta:v1.TypeMeta{Kind:\"\", APIVersion:\"\"}, ObjectMeta:v1.ObjectMeta{Name:\"cancel-me.163d53abc302932f\", GenerateName:\"\", Namespace:\"arendelle-khkkw\", SelfLink:\"/api/v1/namespaces/arendelle-khkkw/events/cancel-me.163d53abc302932f\", UID:\"cd2d7148-1b0d-41a4-b2f7-66d4575c78e0\", ResourceVersion:\"423\", Generation:0, CreationTimestamp:v1.Time{Time:time.Time{wall:0x0, ext:63738125840, loc:(*time.Location)(0x4215be0)}}, DeletionTimestamp:(*v1.Time)(nil), DeletionGracePeriodSeconds:(*int64)(nil), Labels:map[string]string(nil), Annotations:map[string]string(nil), OwnerReferences:[]v1.OwnerReference(nil), Finalizers:[]string(nil), ClusterName:\"\", ManagedFields:[]v1.ManagedFieldsEntry(nil)}, InvolvedObject:v1.ObjectReference{Kind:\"PipelineRun\", Namespace:\"arendelle-khkkw\", Name:\"cancel-me\", UID:\"766809d5-cc12-4e43-a4dd-2fb0524b4974\", APIVersion:\"tekton.dev/v1beta1\", ResourceVersion:\"2973\", FieldPath:\"\"}, Reason:\"Failed\", Message:\"PipelineRun \\\"cancel-me\\\" was cancelled\", Source:v1.EventSource{Component:\"PipelineRun\", Host:\"\"}, FirstTimestamp:v1.Time{Time:time.Time{wall:0x0, ext:63738125839, loc:(*time.Location)(0x4215be0)}}, LastTimestamp:v1.Time{Time:time.Time{wall:0x0, ext:63738125839, loc:(*time.Location)(0x4215be0)}}, Count:1, Type:\"Warning\", EventTime:v1.MicroTime{Time:time.Time{wall:0x0, ext:0, loc:(*time.Location)(nil)}}, Series:(*v1.EventSeries)(nil), Action:\"\", Related:(*v1.ObjectReference)(nil), ReportingController:\"\", ReportingInstance:\"\"}, &v1.Event{TypeMeta:v1.TypeMeta{Kind:\"\", APIVersion:\"\"}, ObjectMeta:v1.ObjectMeta{Name:\"cancel-me-task-8dk65.163d53abde283563\", GenerateName:\"\", Namespace:\"arendelle-khkkw\", SelfLink:\"/api/v1/namespaces/arendelle-khkkw/events/cancel-me-task-8dk65.163d53abde283563\", UID:\"4f461f1a-61f7-4329-81e8-4eb519d10e5a\", ResourceVersion:\"428\", Generation:0, CreationTimestamp:v1.Time{Time:time.Time{wall:0x0, ext:63738125840, loc:(*time.Location)(0x4215be0)}}, DeletionTimestamp:(*v1.Time)(nil), DeletionGracePeriodSeconds:(*int64)(nil), Labels:map[string]string(nil), Annotations:map[string]string(nil), OwnerReferences:[]v1.OwnerReference(nil), Finalizers:[]string(nil), ClusterName:\"\", ManagedFields:[]v1.ManagedFieldsEntry(nil)}, InvolvedObject:v1.ObjectReference{Kind:\"TaskRun\", Namespace:\"arendelle-khkkw\", Name:\"cancel-me-task-8dk65\", UID:\"c2245e73-bde6-4ed5-9b31-f0e679a205c5\", APIVersion:\"tekton.dev/v1beta1\", ResourceVersion:\"2977\", FieldPath:\"\"}, Reason:\"Failed\", Message:\"TaskRun \\\"cancel-me-task-8dk65\\\" was cancelled\", Source:v1.EventSource{Component:\"TaskRun\", Host:\"\"}, FirstTimestamp:v1.Time{Time:time.Time{wall:0x0, ext:63738125840, loc:(*time.Location)(0x4215be0)}}, LastTimestamp:v1.Time{Time:time.Time{wall:0x0, ext:63738125840, loc:(*time.Location)(0x4215be0)}}, Count:1, Type:\"Warning\", EventTime:v1.MicroTime{Time:time.Time{wall:0x0, ext:0, loc:(*time.Location)(nil)}}, Series:(*v1.EventSeries)(nil), Action:\"\", Related:(*v1.ObjectReference)(nil), ReportingController:\"\", ReportingInstance:\"\"}, &v1.Event{TypeMeta:v1.TypeMeta{Kind:\"\", APIVersion:\"\"}, ObjectMeta:v1.ObjectMeta{Name:\"cancel-me-task-8dk65.163d53abde283563\", GenerateName:\"\", Namespace:\"arendelle-khkkw\", SelfLink:\"/api/v1/namespaces/arendelle-khkkw/events/cancel-me-task-8dk65.163d53abde283563\", UID:\"4f461f1a-61f7-4329-81e8-4eb519d10e5a\", ResourceVersion:\"431\", Generation:0, CreationTimestamp:v1.Time{Time:time.Time{wall:0x0, ext:63738125840, loc:(*time.Location)(0x4215be0)}}, DeletionTimestamp:(*v1.Time)(nil), DeletionGracePeriodSeconds:(*int64)(nil), Labels:map[string]string(nil), Annotations:map[string]string(nil), OwnerReferences:[]v1.OwnerReference(nil), Finalizers:[]string(nil), ClusterName:\"\", ManagedFields:[]v1.ManagedFieldsEntry(nil)}, InvolvedObject:v1.ObjectReference{Kind:\"TaskRun\", Namespace:\"arendelle-khkkw\", Name:\"cancel-me-
task-8dk65\", UID:\"c2245e73-bde6-4ed5-9b31-f0e679a205c5\", APIVersion:\"tekton.dev/v1beta1\", ResourceVersion:\"2977\", FieldPath:\"\"}, Reason:\"Failed\", Message:\"TaskRun \\\"cancel-me-task-8dk65\\\" was cancelled\", Source:v1.EventSource{Component:\"TaskRun\", Host:\"\"}, FirstTimestamp:v1.Time{Time:time.Time{wall:0x0, ext:63738125840, loc:(*time.Location)(0x4215be0)}}, LastTimestamp:v1.Time{Time:time.Time{wall:0x0, ext:63738125840, loc:(*time.Location)(0x4215be0)}}, Count:2, Type:\"Warning\", EventTime:v1.MicroTime{Time:time.Time{wall:0x0, ext:0, loc:(*time.Location)(nil)}}, Series:(*v1.EventSeries)(nil), Action:\"\", Related:(*v1.ObjectReference)(nil), ReportingController:\"\", ReportingInstance:\"\"}"
    TestTaskRunPipelineRunCancel/retries=0: init_test.go:74: ############################################
    TestTaskRunPipelineRunCancel/retries=0: init_test.go:75: ### Dumping objects from arendelle-khkkw ###
    TestTaskRunPipelineRunCancel/retries=0: init_test.go:76: ############################################
    TestTaskRunPipelineRunCancel/retries=0: panic.go:634: 
        ---
        apiVersion: tekton.dev/v1beta1
        kind: PipelineRun
        metadata:
          creationTimestamp: "2020-10-12T18:57:16Z"
          generation: 2
          labels:
            tekton.dev/pipeline: cancel-me
          name: cancel-me
          namespace: arendelle-khkkw
          resourceVersion: "3015"
          selfLink: /apis/tekton.dev/v1beta1/namespaces/arendelle-khkkw/pipelineruns/cancel-me
          uid: 766809d5-cc12-4e43-a4dd-2fb0524b4974
        spec:
          pipelineSpec:
            tasks:
            - name: task
              retries: 1
              taskSpec:
                metadata: {}
                steps:
                - image: busybox
                  name: ""
                  resources: {}
                  script: sleep 5000
          serviceAccountName: default
          status: PipelineRunCancelled
          timeout: 1h0m0s
        status:
          completionTime: "2020-10-12T18:57:19Z"
          conditions:
          - lastTransitionTime: "2020-10-12T18:57:19Z"
            message: PipelineRun "cancel-me" was cancelled
            reason: PipelineRunCancelled
            status: "False"
            type: Succeeded
          pipelineSpec:
            tasks:
            - name: task
              retries: 1
              taskSpec:
                metadata: {}
                steps:
                - image: busybox
                  name: ""
                  resources: {}
                  script: sleep 5000
          startTime: "2020-10-12T18:57:16Z"
          taskRuns:
            cancel-me-task-8dk65:
              pipelineTaskName: task
              status:
                completionTime: "2020-10-12T18:57:20Z"
                conditions:
                - lastTransitionTime: "2020-10-12T18:57:20Z"
                  message: TaskRun "cancel-me-task-8dk65" was cancelled
                  reason: TaskRunCancelled
                  status: "False"
                  type: Succeeded
                podName: cancel-me-task-8dk65-pod-v8cxb
                startTime: "2020-10-12T18:57:16Z"
                steps:
                - container: step-unnamed-0
                  name: unnamed-0
                  terminated:
                    exitCode: 1
                    finishedAt: "2020-10-12T18:57:20Z"
                    reason: TaskRunCancelled
                    startedAt: null
                taskSpec:
                  steps:
                  - image: busybox
                    name: ""
                    resources: {}
                    script: sleep 5000
        
        ---
        apiVersion: tekton.dev/v1beta1
        kind: TaskRun
        metadata:
          annotations:
            pipeline.tekton.dev/release: devel
          creationTimestamp: "2020-10-12T18:57:16Z"
          generation: 2
          labels:
            app.kubernetes.io/managed-by: tekton-pipelines
            tekton.dev/pipeline: cancel-me
            tekton.dev/pipelineRun: cancel-me
            tekton.dev/pipelineTask: task
          name: cancel-me-task-8dk65
          namespace: arendelle-khkkw
          ownerReferences:
          - apiVersion: tekton.dev/v1beta1
            blockOwnerDeletion: true
            controller: true
            kind: PipelineRun
            name: cancel-me
            uid: 766809d5-cc12-4e43-a4dd-2fb0524b4974
          resourceVersion: "3010"
          selfLink: /apis/tekton.dev/v1beta1/namespaces/arendelle-khkkw/taskruns/cancel-me-task-8dk65
          uid: c2245e73-bde6-4ed5-9b31-f0e679a205c5
        spec:
          resources: {}
          serviceAccountName: default
          status: TaskRunCancelled
          taskSpec:
            steps:
            - image: busybox
              name: ""
              resources: {}
              script: sleep 5000
          timeout: 1h0m0s
        status:
          completionTime: "2020-10-12T18:57:20Z"
          conditions:
          - lastTransitionTime: "2020-10-12T18:57:20Z"
            message: TaskRun "cancel-me-task-8dk65" was cancelled
            reason: TaskRunCancelled
            status: "False"
            type: Succeeded
          podName: cancel-me-task-8dk65-pod-v8cxb
          startTime: "2020-10-12T18:57:16Z"
          steps:
          - container: step-unnamed-0
            name: unnamed-0
            terminated:
              exitCode: 1
              finishedAt: "2020-10-12T18:57:20Z"
              reason: TaskRunCancelled
              startedAt: null
          taskSpec:
            steps:
            - image: busybox
              name: ""
              resources: {}
              script: sleep 5000
        
    TestTaskRunPipelineRunCancel/retries=0: init_test.go:74: #####################################################
    TestTaskRunPipelineRunCancel/retries=0: init_test.go:75: ### Dumping logs from Pods in the arendelle-khkkw ###
    TestTaskRunPipelineRunCancel/retries=0: init_test.go:76: #####################################################
    TestTaskRunPipelineRunCancel/retries=0: build_logs.go:34: Could not get logs for pod cancel-me-task-8dk65-pod-v8cxb: pods "cancel-me-task-8dk65-pod-v8cxb" not found
```

### PR DESCRIPTION
**Pull request by afrittoli**
# Changes

Fixes #3374

Signed-off-by: Andrea Frittoli <andrea.frittoli@uk.ibm.com>

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [ ] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [ ] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [ ] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)
- [ ] Release notes block has been filled in or deleted (only if no user facing changes)

_See [the contribution guide](https://github.com/tektoncd/pipeline/blob/master/CONTRIBUTING.md) for more details._

Double check this list of stuff that's easy to miss:

- If you are adding [a new binary/image to the `cmd` dir](../cmd), please update
  [the release Task](../tekton/publish.yaml) to build and release this image.

## Reviewer Notes

If [API changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md) are included, [additive changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#additive-changes) must be approved by at least two [OWNERS](https://github.com/tektoncd/pipeline/blob/master/OWNERS) and [backwards incompatible changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#backwards-incompatible-changes) must be approved by [more than 50% of the OWNERS](https://github.com/tektoncd/pipeline/blob/master/OWNERS), and they must first be added [in a backwards compatible way](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#backwards-compatible-changes-first).

# Release Notes

### CODE REVIEW

**Review by sbwsg — APPROVED** [review]

File: test/cancel_test.go [hunk 1]
```diff
@@ -144,23 +144,41 @@ func TestTaskRunPipelineRunCancel(t *testing.T) {
...
+			matchKinds := map[string][]string{"PipelineRun": {pipelineRunName}}
+			// Expected failure events: 1 for the pipelinerun cancel
+			expectedNumberOfEvents := 1
 			t.Logf("Making sure %d events were created from pipelinerun with kinds %v", expectedNumberOfEvents, matchKinds)
 			events, err := collectMatchingEvents(ctx, c.KubeClient, namespace, matchKinds, "Failed")
 			if err != nil {
 				t.Fatalf("Failed to collect matching events: %q", err)
 			}
-			if len(events) != expectedNumberOfEvents {
+			if len(events) < expectedNumberOfEvents {
```
Comment (pritidesai) [inline test/cancel_test.go:155]:
```
grep -lr "expectedNumberOfEvents" test/
test//pipelinerun_test.go
test//v1alpha1/pipelinerun_test.go
test//v1alpha1/cancel_test.go
test//cancel_test.go
```

Various `e2e` tests are validating events (just searched a common variable). The same failure might reappear in any of these tests. We are fixing `cancel` tests here. How about `v1alpha1` cancel tests? I dont remember if I saw it in `v1alpha1`.

We are fixing tests here but the underlying logic still remains as is. Might need fix in future, until then `lgtm`

/lgtm

Comment (afrittoli) [reply test/cancel_test.go:155]: Yeah, this just patches up a frequent failure, but in fact I don't understand why the cancel test started failing all of a sudden, and why only the beta1!

I filed https://github.com/tektoncd/pipeline/issues/3381 to keep track of the remaining work needed.

Comment (pritidesai) [reply test/cancel_test.go:155]: yup thats a mystery 🕵️ , thanks for creating a followup FR 👍

Comment (pritidesai) [reply test/cancel_test.go:155]: Experienced the same failure in the `v1alpha1` 😢, see [logs](https://tekton-releases.appspot.com/build/tekton-prow/pr-logs/pull/tektoncd_pipeline/3068/pull-tekton-pipeline-integration-tests/1316238489435508737/#testv1alpha1-testtaskrunpipelineruncancelretries1)

Comment (afrittoli) [reply test/cancel_test.go:155]: I fixed the `v1alpha1` too now

File: test/cancel_test.go [hunk 2]
```diff
@@ -144,23 +144,41 @@ func TestTaskRunPipelineRunCancel(t *testing.T) {
...
+				t.Fatalf("Failed to collect matching events: %q", err)
+			}
+			if len(events) < expectedNumberOfEvents {
+				collectedEvents := ""
+				for i, event := range events {
+					collectedEvents += fmt.Sprintf("%#v", event)
+					if i < (len(events) - 1) {
+						collectedEvents += ", "
+					}
+				}
```
Comment (mattmoor) [inline test/cancel_test.go:180]:
```suggestion
				collectedEvents := make([]string, 0, len(events))
				for _, event := range events {
					collectedEvents = append(collectedEvents, fmt.Sprintf("%#v", event))
				}
				... = strings.Join(collectedEvents, ", ")
```

**Comment by afrittoli** [comment]
@pritidesai @sbwsg it would be good to get this merged, as it keeps causing flakes

**Review by mattmoor — COMMENTED** [review]
/lgtm
/hold

hold for the nit, if you care, but otherwise I'd love to harden the test and reduce the noise I see from this.

File: test/cancel_test.go (hunk 2 above)
Comment (afrittoli) [reply test/cancel_test.go]: Thanks, done

**Review by pritidesai — APPROVED** [review]
thank you @afrittoli!

**Comment by afrittoli** [comment]
workspace in sidecar might be the next flake?

**Comment by mattmoor** [comment]
/hold cancel

As well.

**Pull request merged**